client-libraries/rust/src/plugin_utils/ingest.rs

```diff
@@ -0,0 +1,1065 @@
+        timeline_attrs: impl IntoIterator<Item = (&str, AttrVal)>,
+    ) -> Result<(), Box<dyn std::error::Error>> {
+        let mut interned_attrs =
+            vec![(self.prep_timeline_attr("timeline.name").await?, name.into())];
+
+        if let Some(run_id) = self.run_id.clone() {
+            let k = self.prep_timeline_attr("timeline.run_id").await?;
+            interned_attrs.push((k, AttrVal::String(run_id.into())));
+        }
+
+        if let Some(time_domain) = self.time_domain.clone() {
+            let k = self.prep_timeline_attr("timeline.time_domain").await?;
+            interned_attrs.push((k, AttrVal::String(time_domain.into())));
+        }
+
+        interned_attrs.extend(self.additional_timeline_attributes.iter().cloned());
+        interned_attrs.extend(self.override_timeline_attributes.iter().cloned());
+
+        for (k, v) in timeline_attrs {
+            let k = self.prep_timeline_attr(k).await?;
+            if self
+                .override_timeline_attributes
+                .iter()
+                .any(|(ko, _)| k == *ko)
+            {
+                continue;
+            }
+
+            interned_attrs.push((k, v));
+        }
+
+        self.inner.timeline_metadata(interned_attrs).await?;
+
+        Ok(())
+    }
+
+    async fn prep_timeline_attr(
+        &mut self,
+        k: &str,
+    ) -> Result<InternedAttrKey, Box<dyn std::error::Error>> {
+        let key = normalize_timeline_key(k);
+        let int_key = if let Some(ik) = self.timeline_keys.get(&key) {
+            *ik
+        } else {
+            let ik = self.inner.declare_attr_key(key.clone()).await?;
+            self.timeline_keys.insert(key, ik);
+            ik
+        };
+
+        Ok(int_key)
+    }
+
+    /// Create an event on the current timeline.
+    ///
+    /// <div class="warning">`switch_timeline` must be called at least once before calling `send_event`! </div>
+    ///
+    /// * `name`: The event name; sets the `event.name` attr.
+    ///
+    /// * `ordering`: The relative ordering of this event on its timeline. Most users will use a local counter
+    ///   to populate this value, and increment it each time an event is sent on the timeline. Values to not
+    ///   have to be consecutive, so you can use a single counter that is shared between all timelines if that makes
+    ///   sense for your application.
+    ///
+    ///   <div class="warning">Avoid sending duplicate `ordering` values for the same timeline.</div>
+    ///
+    /// * `attrs`: The attributes to attach to the event. While you can use this with anything
+    ///   that implements `IntoIterator`, it's idiomatic to use a literal slice, and to
+    ///   use `into()` to convert values to `AttrVal`:
+    ///   `client.send_event("ev", [("attr1", 42.into())]).await?;`
+    ///
+    ///   * These keys are automatically normalized, so you prepending "event." is optional.
+    ///
+    ///   * If "timestamp" or "event.timestamp" is not given here, the
```

Comment:
I think this is useful behavior to have, but worth noting that some plugins will intentionally not have timestamps and we may want to make an opt-in/out switch for this.